protocol/contracts/beanstalk/farm/TokenFacet.sol

```diff
@@ -81,7 +82,7 @@ contract TokenFacet is IERC1155Receiver, ReentrancyGuard {
         address recipient,
         uint256 amount,
         LibTransfer.To toMode
-    ) external payable nonReentrant {
+    ) external payable fundsSafu noSupplyChange nonReentrant {
```

Comment:
Many of the external functions that transfer an amount specify an amount that is brought in/out of the system. What are your thoughts on an invariant that verifies that delta is held true? For example, if `toMode` is `internal`, we can enforce `noNetFlow`. If external, we can verify that 1) token balance has changed at minimum `amount`  , and that one of the internal parameters increased by that amount.

Reply:
🤔🤔🤔🤔🤔

Reply:
i looked into this and implemented an invariant that limits outflow to a single asset and puts a cap on the amount. however, when distributing it across the code it became clear that it is not very common to know the amount at function call time, and even when it is known it is just an argument set by the caller. i feel there was limited value in verifying values dictated by the theoretical attacker.

instead i pivoted to two simple new invariants `noOutFlow` and `oneOutFlow` which restrict the movement of asset to either be all inbound or limited to a single token out, which is typically inferable from the action.

these three variations of the invariants can cover everything except chop and convert (convert may be compatible on pipeline convert).

Reply:
ref:

https://github.com/BeanstalkFarms/Beanstalk/pull/821/commits/5ca8c766c339746a3e512dfbcec3b5241d11e5eb#diff-ec8cf17e4466d63fc908daec7f18d12a489cc75bd0ac653a63305ef21faddc99L183

https://github.com/BeanstalkFarms/Beanstalk/pull/821/commits/5ca8c766c339746a3e512dfbcec3b5241d11e5eb#diff-ec8cf17e4466d63fc908daec7f18d12a489cc75bd0ac653a63305ef21faddc99L191

Reply:
lgtm